forbidigo: better support for configuring complex rules

pohly opened the pull request:
forbidigo 1.4.0 introduced a new configuration mechanism with multiple fields per rule. It still supports a single string, but allowing structs as alternative to such strings makes the configuration easier to read and write.

A new global flag is needed to enable the more advanced analysis.

ashanbrown commented on .golangci.reference.yml:
```diff
@@ -358,14 +358,30 @@ linters-settings:
+    # methods and fields) the type name. This makes it possible to handle
+    # import renaming and forbid struct fields and methods.
+    # Default: false
+    expand-expressions: true
```
I think this is meant to be called `analyze-types`, which is the name of the analyzer flag.

pohly replied:
Right. I forget to update this after we renamed that flag. Fixed.

ashanbrown commented on pkg/golinters/forbidigo.go:
```diff
@@ -40,14 +41,20 @@ func NewForbidigo(settings *config.ForbidigoSettings) *goanalysis.Linter {
 	}
 
+	loadMode := goanalysis.LoadModeSyntax
+	if settings != nil && settings.ExpandExpressions {
```
I believe we also need to set the analyzer flag called `analyze_types` when we see this set.

pohly replied:
Which analyzer flag?

goanalysis uses different "levels", not individual flags: https://pkg.go.dev/github.com/golangci/golangci-lint/pkg/golinters/goanalysis#LoadMode

ashanbrown replied:
I was thinking we were using forbidigo as an analyzer, per the new guidelines for linters.  For an analyzer, we'd set the `analyze_types` flag at https://github.com/ashanbrown/forbidigo/blob/master/pkg/analyzer/analyzer.go#L55.  That said, it looks like we've set up forbidigo to automatically expand types when `TypesInfo` is provided. I think I hadn't quite realized we were using that as a flag, rather than just information.  In retrospect, my preference would have been to explicitly add a new `analyzeTypes` option at https://github.com/ashanbrown/forbidigo/blob/master/forbidigo/forbidigo.go#L65 (similar for what we did for the analyzer mode).

ldez replied:
By using types, your linter is now considered a slow linter.
https://github.com/golangci/golangci-lint/pull/3639

ashanbrown replied:
@ldez That makes sense, and we had anticipated this issue, although I didn't realize that it was formally checked in golangci-lint.  The linter can work without types, but there is an opt-in behavior that allows us to use types to match more things.  Is there any precedent for a linter being both fast and slow?  I don't suppose we could have both fast and slow versions of the same linter?  I'd imagine they would have to have separate names and configurations.

ldez replied:
A linter cannot be slow and fast, because the slow part is the type analysis, and the type analysis is not an option handled by the linter itself.
I don't think that duplicating a linter is a solution.
Your linter will just be "slow" now.

ashanbrown replied:
Is type analysis triggered by the line at https://github.com/golangci/golangci-lint/pull/3612/files#diff-b647b23fb0e6073b7f161d88a8cd16af6dd3e90de1d6a81006d2d9a4ee416707R424?  In this PR, it appears that setting is conditionally executed based on that configuration.

pohly replied:
The goal was to not change behavior unless the user explicitly asks  for it in the config. It looks like there are two places where that is relevant:
- GetAllSupportedLinterConfigs: covered by this PR
- NewForbidigo: covered by this PR

The already [merged commit](https://github.com/golangci/golangci-lint/pull/3639/commits/06b681e96329e96779b008c301bc5f9460b6c111) unconditionally reconfigures forbidigo to enable type analysis. There's a certain risk that this breaks users because existing patterns get interpreted differently. I missed earlier that the PR wasn't just the automatic bump of the forbidigo package.

I think we should merge this PR here soon because it is required to finish the forbidigo update to 1.5.x properly.

pohly replied:
Wait, I'm not keeping up with you guys :cry: 

So forbidigo 1.5.x is new, and the upgrade of it in https://github.com/golangci/golangci-lint/pull/3639 is something that I really only saw today because it's also new.

pohly replied:
That PR makes forbidigo a "slow" linter, without actually enabling the feature which uses the additional information (no call of `OptionAnalyzeTypes(true)`.

So my statement above still stands: this PR is needed to finish the forbidigo update. I just need to add `OptionAnalyzeTypes(true)`.

pohly replied:
I've added it.

ashanbrown commented:
FYI, I've bumped `forbidigo` to v1.5.0 which now requires a new option called `OptionAnalyzeTypes` to be set to enable analyzing types.  I'm also looking at integrating one more change to remove pkg/errors, so it might not hurt to upgrade all that together.  That will be version v1.5.1.

ashanbrown commented on .golangci.reference.yml:
```diff
@@ -361,14 +361,31 @@ linters-settings:
-      # Optionally put comments at the end of the regex, surrounded by `(# )?`
-      # Escape any special characters.
+      # Optional message that gets included in error reports.
+      - pattern: ^fmt\.Print.*$
```
I think we ended up calling this "p" at https://github.com/golangci/golangci-lint/pull/3612/files#diff-cb1fbdbfccd3fd0d8fe9d9006138fbcd1afcc3a701b37180c6fe50ed0c5c390fR332.

pohly replied:
True, and I did the same for `ForbidigoPattern` in `linter_settings.go`. I've fixed the reference file to match.

It's worth calling out that we can use `pattern` here if `p` is considered too brief: we can map that in `linter_settings.go`:
```
// ForbidigoPattern corresponds to forbidigo.pattern and adds
// mapstructure support. The YAML field names must match what
// forbidigo expects.
type ForbidigoPattern struct {
	// patternString gets populated when the config contains a string as
	// entry in ForbidigoSettings.Forbid[] because ForbidigoPattern
	// implements encoding.TextUnmarshaler and the reader uses the
	// mapstructure.TextUnmarshallerHookFunc as decoder hook.
	//
	// If the entry is a map, then the other fields are set as usual
	// by mapstructure.
	patternString string

	Pattern string `yaml:"p" mapstructure:"pattern"` // <=====
	Package string `yaml:"pkg,omitempty" mapstructure:"pkg,omitempty"`
	Msg     string `yaml:"msg,omitempty" mapstructure:"msg,omitempty"`
}
```

I myself think golangci-lint should follow the choice made for upstream forbidigo and use `p`, too.

ashanbrown commented on pkg/lint/lintersdb/manager.go:
```diff
@@ -414,10 +414,17 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
+				WithURL("https://github.com/ashanbrown/forbidigo")
+			if forbidigoCfg != nil && forbidigoCfg.AnalyzeTypes {
+				// Need more information for the analysis.
+				l = l.WithLoadForGoAnalysis()
```
@ldez As @pohly mentioned, we are making this call (which sets `IsSlow`) conditional on the configuration.  Are you ok with having this otherwise "fast" linter become slow if this flag is set or would that be considered too unexpected?

ldez replied:
`GetAllSupportedLinterConfigs` is called in different situations, the linter config will not be available in all those situations, so please remove this and don't try to create an asymetrical configuration between `WithLoadForGoAnalysis` and `LoadModeTypesInfo`.

pohly replied:
I see how not setting `IsSlow` in `GetAllSupportedLinterConfigs` when called without config is wrong (or at least potentially misleading).

But is it also useful in `NewForbidigo` to use `LoadModeTypesInfo` whether it's needed or not? Can something break when combining `WithLoadForGoAnalysis` with `LoadModeSyntax`?

pohly replied:
Okay. I've pushed an update which always uses the more expensive mode. In the long run I'd expect users to switch to it anyway. It's just not the default because of some risk of breaking certain patterns.

ashanbrown replied:
For the record, I'm not sure I agree, and I have to admit I'm now starting to regret my decision to allow type-based checking.  I think forbidigo is most useful if it can be run as a pre-commit check, and my guess is that this outweighs the value of being able to check these additional kinds of patterns (which still behave unexpectedly in a few cases).  I would strongly favor finding some way to make this linter behave slowly only if the `analyzeTypes` config is set.

pohly replied:
Is anyone really running only "fast" linters in a pre-commit check and the "slow" ones after merging? That seems odd because then mistakes get merged and need to be fixed after the fact.

I can see the advantage of running only "fast" linters while developing the code to speed up the write/compile/test loop, but I am not sure how important it is to include forbidigo in that loop.

ashanbrown replied:
I don't think the pattern of running a fast set of checks on pre-commit is too unusual.  It's one I've employed frequently.  As the most extreme example, you could likely run all your tests on pre-commit, but we tend to do that only in CI because they're too slow.  The example cases of `fmt.Println` and `ginkgo.FDescribe` are the kinds of thing that are great to catch on pre-commit.  It's nice not to have to wait for a whole CI cycle to know that you left in some debug code.

ldez replied:
pre-commit is not something so usual because pre-commit has several problems (cannot be committed, require an installation of the hook, etc.).
Only running "fast" linters and using the CI to run "slow" linters is also not so usual.

Personally and professionally, I don't use pre-commit, and I don't want to wait for a CI to get feedback from the linter (CIs are slower than my computer).

pohly replied:
Now I think I understand what @ashanbrown meant with "pre-commit". I don't force linter runs when committing code either. Locally, I run linters when I am ready to submit something (unless I forget...) or when I believe that it might find something.

In Kubernetes, we run all linters when testing a pull request, and those results are merge-blocking, i.e. we never merge code that has linter issues. That's what I meant earlier.

ashanbrown replied:
Back to the main point of this discussion: @ldez, you had asked that we not be asymmetric in our use of `WithLoadForGoAnalysis` in one place and not the other.  Could we simply report that we need the requirements of a fast linter except when this "experimental" setting is set in the config file?  When we see the experimental config, we'd add the extra load requirements.  I realize this is still asymmetric, although at least it would no longer call `WithLoadForGoAnalysis`.  Is the concern that by reporting a linter as fast, say in a help message, and later making it slow, we would be confusing the user, or is there something that would actually break?

pohly replied:
`go test -v -run TestSourcesFromTestdata/forbidigo ./test` doesn't pass when `GetAllSupportedLinterConfigs` doesn't use `WithLoadForGoAnalysis` for forbidigo, so it has technical implications.

If we absolutely need a "fast forbidigo" (not convinced yet myself, but usage patterns are different and I don't want to make forbidigo less useful for its original author!) and a "throrough forbidigo" (a strong yes from me for Kubernetes, because import aliases make the current source-code based rules unreliable), then we could register forbidigo twice:
- Once with the current configuration and mode of operation (no changes).
- Again with a different name (working title "forbidigo-analyze-types", better suggestions welcome) and the new configuration without the `AnalyzeTypes` field.

ashanbrown replied:
I'm not sure it's `LoadModeTypesInfo` that is causing us to be considered a slow linter.  For example, durationCheck uses this load mode and is still considered fast.  I think it's the requirement to load dependencies that would make this a slow linter.  `WithLoadForGoAnalysis` does this:

```
lc.LoadMode |= packages.NeedImports | packages.NeedDeps | packages.NeedExportFile | packages.NeedTypesSizes
```

This was added at https://github.com/golangci/golangci-lint/pull/1844.  Perhaps @ldez can fill us on as to why we can't just set `IsSlow` based on the presence of select LoadMode flags.

I'd personally be happy with the multiple-linter configs but that had sounded like a non-starter to me.

ldez replied:
`goanalysis.LoadModeTypesInfo` required `WithLoadForGoAnalysis()`.

A linter with two "speed states" doesn't seem right.

"slow" or "fast" don't really describe the speed of a linter, it's just a best-effort state.
The "slow" linters share a state for the slow part, so if you run only one linter you will pay the cost but if your run several linters the cost will be shared.

Duplicating the linter doesn't seem like the right idea too.

I think it's not really important if your linter is flagged as "slow" or "fast".
But if it's important to you, you have to make choice inside the linter: use types or not.

I don't want to create unneeded complexity inside the configuration, I also prefer maintainability over possibilities.

pohly replied:
Would it work to register it with `WithLoadForGoAnalysis` unconditionally and then decide in `NewForbidigo` based on the actual setting? At least the tests are passing that way.

Would it make the linter run faster? I'm not so sure.

For Kubernetes, `golangci-lint run --disable-all --enable forbidigo -v` with that approach uses `LoadModeSyntax`:
```
INFO Execution took 8.982228038s
```

That's after cleaning the golangci-lint cache, but with a warm Go build cache.

Without this hack it's 9.556500684s - a bit slower, but not much.

With an empty Go cache (`go clean --cache`), the difference is 56.58711488s (with that approach) vs. 57.399810283s (without).

For comparison, with forbidigo master (985efcc1) it's 9.639382845s and 57.23967985s.

These numbers are so close together, I don't see a relevant difference. It probably comes down to noise - otherwise master would have to be the fastest.

@ashanbrown: can you give a specific example where you want to use "fast" analysis? Does this PR really make a significant difference?

pohly commented:
My latest push triggered some linter warnings. I've fixed those:
```patch
diff --git a/pkg/golinters/forbidigo.go b/pkg/golinters/forbidigo.go
index cad20b4e..b46c8698 100644
--- a/pkg/golinters/forbidigo.go
+++ b/pkg/golinters/forbidigo.go
@@ -16,6 +16,7 @@ import (
 
 const forbidigoName = "forbidigo"
 
+//nolint:dupl
 func NewForbidigo(settings *config.ForbidigoSettings) *goanalysis.Linter {
        var mu sync.Mutex
        var resIssues []goanalysis.Issue
@@ -59,9 +60,7 @@ func runForbidigo(pass *analysis.Pass, settings *config.ForbidigoSettings) ([]go
                forbidigo.OptionExcludeGodocExamples(settings.ExcludeGodocExamples),
                // disable "//permit" directives so only "//nolint" directives matters within golangci-lint
                forbidigo.OptionIgnorePermitDirectives(true),
-       }
-       if settings != nil && settings.AnalyzeTypes {
-               options = append(options, forbidigo.OptionAnalyzeTypes(true))
+               forbidigo.OptionAnalyzeTypes(settings.AnalyzeTypes),
        }
 
        // Convert patterns back to strings because that is what NewLinter
```

It looks like `runForbidigo` will never be called with a nil pointer (existing code already relied on that), so I removed the if check.

alexandear commented on pkg/lint/lintersdb/manager.go:
```diff
@@ -417,6 +417,13 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 		linter.NewConfig(golinters.NewForbidigo(forbidigoCfg)).
 			WithSince("v1.34.0").
 			WithPresets(linter.PresetStyle).
+			// Strictly speaking, the additional
+			// information is only needed when
+			// forbidigoCfg.AnalyzeTypes is chosen by the
+			// user. But we don't know that here in all
+			// cases (sometimes config is not loaded), so
+			// we have to assume that it is needed to be on
+			// the safe side.
```
minor: I'm curious why this comment spreads over 7 lines of code. I think it can get in 3 lines (because of `lll = 140`)
```suggestion
			// Strictly speaking, the additional information is only needed when forbidigoCfg.AnalyzeTypes
			// is chosen by the user. But we don't know that here in all cases (sometimes config is not loaded),
			// so we have to assume that it is needed to be on the safe side.
```

pohly replied:
I am wrapping at a column that typically is preferred by the various projects that I work in. If golangci-lint accepts longer lines or it simply fits better into the code because it already has longer lines, then I can certainly wrap differently.

ldez replied:
I prefer https://sembr.org/

ldez replied:
4 lines

```go
// Strictly speaking,
// the additional information is only needed when forbidigoCfg.AnalyzeTypes is chosen by the user.
// But we don't know that here in all cases (sometimes config is not loaded),
// so we have to assume that it is needed to be on the safe side.
```

pohly replied:
Thanks. I've updated this comment and also some others to match this style.

sbrownjc commented:
Looking forward to this getting merged! Once it is, please update the Schema stored here: https://github.com/SchemaStore/schemastore/blob/e321c0e165562a845e3d05b28f4ce2054bc11d7e/src/schemas/json/golangci-lint.json#L911-L913

ashanbrown commented:
@pohly Could you let us know where this PR is at?  I'm still not excited about this never being a fast linter (although I think that change went in inadvertently with the dependency bump), but I figure it's better that we push it out and get feedback, given that there seems to be interest in the new features.  Thanks.

pohly commented:
@ashanbrown: I think @ldez is just waiting for your okay to merge it as it is now. You were the one who had concerns and it wouldn't be fair to just ignore that.

The discussion in https://github.com/golangci/golangci-lint/pull/3612#discussion_r1126076075 stopped with me asking about some specific example where I can try out myself what the actual difference is. If you feel that this is not necessary, then we can also just go ahead and merge.

alexandear commented on pkg/config/linters_settings.go:
```diff
@@ -330,10 +333,46 @@ type ExhaustructSettings struct {
+
+// String is intentionally not called MarshalText because that led to infinite
+// recursion when yaml.Marshal called MarshalText.
+func (p *ForbidigoPattern) String() ([]byte, error) {
```
`String()` usually returns only `string`. Maybe we should change this function.

ldez reviewed (approved):
LGTM
